airbyte-cdk/java/airbyte-cdk/db-sources/src/main/kotlin/io/airbyte/cdk/integrations/source/relationaldb/streamstatus/TransientErrorTraceEmitterIterator.kt

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.cdk.integrations.source.relationaldb.streamstatus
+
+import io.airbyte.cdk.integrations.base.AirbyteTraceMessageUtility.makeErrorTraceAirbyteMessage
+import io.airbyte.commons.util.AutoCloseableIterator
+import io.airbyte.protocol.models.v0.AirbyteErrorTraceMessage
+import io.airbyte.protocol.models.v0.AirbyteMessage
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+
+class TransientErrorTraceEmitterIterator(private val e: Throwable) :
+    AutoCloseableIterator<AirbyteMessage?> {
+    private var emitted = false
+
+    override fun hasNext(): Boolean {
+        return !emitted
+    }
+
+    override fun next(): AirbyteMessage {
+        emitted = true
+        return makeErrorTraceAirbyteMessage(
+            e,
+            e.message,
+            AirbyteErrorTraceMessage.FailureType.TRANSIENT_ERROR
+        )
+    }
```

Comment:
How would this iterator getting used?

I imagine this would be like stream status iterator wrapped around the original iterator; sounds like we are doing it with a catch-all try catch block, and if there's exception we will feed it into this one?

And I guess we are not using the original exception handler anymore right?

Reply:
It will be used similar to the iterator here : https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-mysql/src/main/java/io/airbyte/integrations/source/mysql/initialsync/MySqlInitialLoadHandler.java#L105

It will be the last Airbyte message emitted (after the last stream complete message)

With the WASS algorithm, we're swapping the order in which CDC iterators and initial snapshot iterators run. One side effect of this is that we are unable to clear the CDC after an iniital sync is completed. Thus, after the initial load is completed we throw a transient error (using the emitter) to force another sync to clear the WAL.